### PR TITLE
ws2812 timings

### PR DESCRIPTION
**Pull request by jacobrosenthal**
Greetings all. Thanks for all the great libraries. 

In attempting to figure out why my atasmd pygamer examples with a perfectly good timer needed to push clock from 3mhz up to 3.1-3.8mhz, and then higher when I removed this seemingly pointless extra for loop that seems it be in most exmaples 
https://github.com/atsamd-rs/atsamd/blob/master/boards/pygamer/examples/neopixel_rainbow_timer.rs#L53

I took a look at the original timings and compared from
https://wp.josh.com/2014/05/13/ws2812-neopixels-are-not-so-finicky-once-you-get-to-know-them/

I believe the current timings are shorting the low part of the 0 and 1 bit. Ive pushed this up by 333ns (1 block) to 666ns for both and Im seeing vastly better results and at 3mhz even, no extra loops, lto etc fudging.

Secondly, the latch timing is much larger than it needs to be, I brought that down to its minimum ~6000ns.

I havent touched the slow path, because I dont understand what numbers thats trying to hit, but I think that needs a look to. Also I realize people tend to try to use these same code paths for other chips where we might want to stretch numbers here to accommodate. However, it would be best if those assumptions were documented.

~~Note this makes all the 'wheel' examples in atsamd not look as pretty, but I think thats because they were built to accommodate the old timings
https://github.com/atsamd-rs/atsamd/blob/master/boards/pygamer/examples/neopixel_rainbow_timer.rs#L55~~

This code transitions colors very nicely so I know its not the drive
```
    loop {
        for j in 0..=255u8 {
            let colors: [RGB8; NUM_LEDS] = [
                wheel(j).into(),
                wheel(j).into(),
                wheel(j).into(),
                wheel(j).into(),
                wheel(j).into(),
            ];

            let _ = neopixel.write(brightness(colors.iter().cloned(), 32));
            delay.delay_ms(5u8);
        }
    }

```

### CODE REVIEW

**Comment by jacobrosenthal** [comment]
Nevermind, wheel examples look same still. I had my brightness too low.

File: src/lib.rs [hunk 1]
```diff
@@ -97,9 +109,11 @@ where
             self.write_byte(item.r);
             self.write_byte(item.b);
         }
-        // Get a timeout period of 300 ns
-        for _ in 0..900 {
-            block!(self.timer.wait()).ok();
+
+        // Latch for > 6000 ns
+        // 18 * 333 = 6327ns
+        for _ in 0..19 {
```
Comment (david-sawatzke) [inline src/lib.rs]: It seems like there are different variations with different reset periods, see https://blog.particle.io/2017/05/11/heads-up-ws2812b-neopixels-are-about-to-change/

Comment (jacobrosenthal) [reply src/lib.rs]: Ah perfect. Thanks for the link, So then the old code was spot on there, but with a bad comment. Ive updated comments and added a link to that particle page as well.

**Comment by jacobrosenthal** [comment]
Ping!

Im starting to get ready to merge code based on this. Ive been using these timings all week and theyve really cleaned up weirdness I was seeing.

Can anyone run this on one of their boards and confirm this still works nice with other possible neopixel ws2812 versions?

Pinging people who own stuff that depends on this crate. (mainly the atsamd org it seems)
@wez @tarcieri @sajattack 

And I still need thoughts on if its necessary or maybe not implement this change for the 'slow' mode

**Comment by sajattack** [comment]
Sorry, I wanted to see what @david-sawatzke thought of this because he is more of the expert here, and  I forgot about this pr.

**Comment by tarcieri** [comment]
@jacobrosenthal I might have time to take a look soon

File: src/lib.rs [hunk 2]
```diff
@@ -61,17 +74,20 @@ where
     fn write_byte(&mut self, mut data: u8) {
         for _ in 0..8 {
             if (data & 0x80) != 0 {
-                block!(self.timer.wait()).ok();
-                self.pin.set_high();
-                block!(self.timer.wait()).ok();
-                block!(self.timer.wait()).ok();
-                self.pin.set_low();
+                //1bit, 666ns on time, 666ns off time
```
Comment (david-sawatzke) [inline src/lib.rs]: The low block was outside done this way, so that the iterator can run during the low time without needing additional overhead

Comment (david-sawatzke) [reply src/lib.rs]: Additionally, doing it your way may cut the time for the high block short (since the timer runs while low)

File: src/lib.rs [hunk 3]
```diff
@@ -8,11 +8,24 @@
...
+
+// But new device reset latch timing is posted as 280us
+// https://blog.particle.io/2017/05/11/heads-up-ws2812b-neopixels-are-about-to-change/
+
 #![no_std]
 
 use embedded_hal as hal;
 
-use crate::hal::digital::OutputPin;
+use crate::hal::digital::v2::OutputPin;
```
Comment (david-sawatzke) [inline src/lib.rs]: Please do this change in a separate pr, we might also want to add (configurable) error handling

**Comment by david-sawatzke** [comment]
Sorry, was a bit busy. Extending the low time seems reasonable, if it helps things. Maybe make it configurable? We should rethink using features for this task, seems a bit hacky.

I'm a bit confused how increasing your timer frequency fixed your problems, since it makes the low period shorter. Could you try with the slow option? It's there so if checking the timer takes too long for the high period of the 0 bit it simply.. doesn't. 

A capture with a logic analyzer would help, since the workaround and the proposed proper fix seem counter to each other.

**Comment by david-sawatzke** [comment]
@sajattack @jacobrosenthal That the extra loop in the example makes a difference is a bit concerning, maybe some weird interaction with the timer periods?

File: src/lib.rs (hunk 3 above)
Comment (david-sawatzke) [reply src/lib.rs]: Done in #2

**Comment by jacobrosenthal** [comment]
So with all the other changes upstreamed, my code seems to be working with the upstream timings. Let me see if it gets weird again over the next week

**Comment by david-sawatzke** [comment]
Maybe some weird interaction with `OldOutputPin`?
But the timings ARE broken for ws2812b, see [this](https://cpldcpu.wordpress.com/2014/01/14/light_ws2812-library-v2-0-part-i-understanding-the-ws2812/), with my (currently) favorite [table](https://cpldcpu.files.wordpress.com/2014/01/ws2812_timing_table.png) for ws2812 timings.

**Comment by jacobrosenthal** [comment]
The upstream timings are working fine still... Im happy to let this sit a bit. 
Would you cut a new release with the pin and ownership changes, it would help me merge and test more in atsamd.

**Comment by sajattack** [comment]
0.3.0 released
